common/libs/transaction.js

```diff
@@ -29,3 +39,95 @@ export type BroadcastTransaction = {|
   rawTx: string,
   signedTx: string
 |};
+
+export async function generateTransaction(
+  node: BaseNode,
+  tx: Transaction,
+  wallet: BaseWallet,
+  token: ?Token
+): Promise<BroadcastTransaction> {
+  // Reject bad addresses
+  if (!isValidETHAddress(tx.to)) {
+    return Promise.reject(new Error(translate('ERROR_5')));
+  }
+
+  // Reject token transactions without data
+  if (token && !tx.data) {
+    return Promise.reject(new Error('Tokens must be sent with data'));
+  }
+
+  // Reject gas limit under 21000 (Minimum for transaction)
+  // Reject if limit over 5000000
+  // TODO: Make this dynamic, the limit shifts
+  const limitBig = new Big(tx.gasLimit);
+  if (limitBig.lessThan(21000)) {
+    return Promise.reject(
+      new Error(translate('Gas limit must be at least 21000 for transactions'))
+    );
+  }
+
+  if (limitBig.greaterThan(5000000)) {
+    return Promise.reject(new Error(translate('GETH_GasLimit')));
+  }
+
+  // Reject gas over 1000gwei (1000000000000)
+  const gasPriceBig = new Big(tx.gasPrice);
+  if (gasPriceBig.greaterThan(new Big('1000000000000'))) {
+    return Promise.reject(
+      new Error(
+        'Gas price too high. Please contact support if this was not a mistake.'
+      )
+    );
+  }
+
+  // Ensure their balance exceeds the amount they're sending
+  // TODO: Include gas price too, tokens should probably check ETH too
```

Comment:
I think we should validate as much as we can. We have no guarantees about future error message from not yet written node software :) This should also skip some of those when creating tx offline. Maybe we should lift/duplicate some checks to confirm/broadcast.